statesync/dispatcher.go

```diff
@@ -0,0 +1,322 @@
+package statesync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/tendermint/tendermint/light/provider"
+	"github.com/tendermint/tendermint/p2p"
+	ssproto "github.com/tendermint/tendermint/proto/tendermint/statesync"
+	proto "github.com/tendermint/tendermint/proto/tendermint/types"
+	"github.com/tendermint/tendermint/types"
+)
+
+var (
+	errNoConnectedPeers    = errors.New("no available peers to dispatch request to")
+	errUnsolicitedResponse = errors.New("unsolicited light block response")
+	errNoResponse          = errors.New("peer failed to respond within timeout")
+	errPeerAlreadyBusy     = errors.New("peer is already processing a request")
+	errDisconnected        = errors.New("dispatcher has been disconnected")
+)
+
+// dispatcher keeps a list of peers and allows concurrent requests for light
+// blocks. NOTE: It is not the responsibility of the dispatcher to verify the
+// light blocks.
+type dispatcher struct {
+	availablePeers *peerlist
+	requestCh      chan<- p2p.Envelope
+	timeout        time.Duration
+
+	mtx     sync.Mutex
+	calls   map[p2p.NodeID]chan *types.LightBlock
+	running bool
+}
+
+func newDispatcher(requestCh chan<- p2p.Envelope, timeout time.Duration) *dispatcher {
+	return &dispatcher{
+		availablePeers: newPeerList(),
+		timeout:        timeout,
+		requestCh:      requestCh,
+		calls:          make(map[p2p.NodeID]chan *types.LightBlock),
+		running:        true,
+	}
+}
+
+func (d *dispatcher) LightBlock(ctx context.Context, height int64) (*types.LightBlock, p2p.NodeID, error) {
+	d.mtx.Lock()
+	outgoingCalls := len(d.calls)
+	d.mtx.Unlock()
```

Comment:
should this be a RW mutex? 

I'm also, always a bit worried about taking mutexes without unlocking in defers in the case of a panic/recover related deadlock. I think we're safe in this case but It's fine?

Reply:
I think it's fine. This is also the only time where we use the lock to make a read. My non-factually backed intuition tells me that a regular Mutex is more efficient than a RWMutex unless you're doing a lot of reads